packages/frontend/core/src/components/affine/page-properties/tags-inline-editor.tsx

```diff
@@ -0,0 +1,381 @@
+            e.stopPropagation();
+            e.preventDefault();
+            updateTagName(e.currentTarget.value);
+          }
+        }}
+        placeholder={t['Untitled']()}
+      />
+    );
+
+    options.push('-');
+
+    options.push({
+      text: t['Delete'](),
+      icon: <DeleteIcon />,
+      type: 'danger',
+      onClick() {
+        legacyProperties.removeTagOption(tag.id);
+      },
+    });
+
+    options.push({
+      text: t['com.affine.page-properties.tags.open-tags-page'](),
+      icon: <TagsIcon />,
+      onClick() {
+        navigate.jumpToTag(legacyProperties.workspaceId, tag.id);
+      },
+    });
+
+    options.push('-');
+
+    options.push(
+      tagColors.map(([name, color]) => {
+        return {
+          text: name,
+          icon: (
+            <div className={styles.tagColorIconWrapper}>
+              <div
+                className={styles.tagColorIcon}
+                style={{
+                  backgroundColor: color,
+                }}
+              />
+            </div>
+          ),
+          checked: tag.color === color,
+          onClick() {
+            legacyProperties.updateTagOption(tag.id, {
+              ...tag,
+              color,
+            });
+          },
+        };
+      })
+    );
+    return renderMenuItemOptions(options);
+  }, [legacyProperties, navigate, t, tag]);
+
+  return (
+    <Menu
+      rootOptions={{
+        modal: true,
+      }}
+      contentOptions={{
+        onClick(e) {
+          e.stopPropagation();
+        },
+      }}
+      items={items}
+    >
+      {children}
+    </Menu>
+  );
+};
+
+export const TagsEditor = ({
+  options,
+  value,
+  onChange,
+  onOptionsChange,
+  readonly,
+}: TagsEditorProps) => {
+  const t = useAFFiNEI18N();
+  const [inputValue, setInputValue] = useState('');
+  const exactMatch = options.find(o => o.value === inputValue);
+  const filteredOptions = useMemo(
+    () =>
+      options.filter(o => (inputValue ? filterOption(o, inputValue) : true)),
+    [inputValue, options]
+  );
+
+  const onInputChange = useCallback(
+    (e: React.ChangeEvent<HTMLInputElement>) => {
+      setInputValue(e.target.value);
+    },
+    []
+  );
+
+  const onAddTag = useCallback(
+    (id: string) => {
+      if (!value.includes(id)) {
+        onChange?.([...value, id]);
+      }
+    },
+    [onChange, value]
+  );
+
+  const [nextColor, rotateNextColor] = useReducer(
+    color => {
+      const idx = tagColors.findIndex(c => c[1] === color);
+      return tagColors[(idx + 1) % tagColors.length][1];
+    },
+    tagColors[Math.floor(Math.random() * tagColors.length)][1]
+  );
+
+  const onCreateTag = useCallback(
+    (name: string) => {
+      if (!name.trim()) {
+        return;
+      }
+
+      const newTag = {
+        id: nanoid(),
+        value: name.trim(),
+        color: nextColor,
+      };
+      rotateNextColor();
+      onOptionsChange?.([...options, newTag]);
+      setTimeout(() => {
```

Comment:
Why do we need a `setTimeout` here? Maybe worth to add a comment to explain

Reply:
It looks like this is not required. Removed